index.rst

```diff
@@ -10,16 +10,200 @@
+This is possible when rebroadcasting the alert stream -- in the worst case, duplicate alerts exist but have the same position and time -- and when running the pipeline from ISR through image differencing.
+However, for reasons explored in :ref:`association`, idempotency cannot be safely achieved with source association.
+
+.. _retries:
+
+Retrying Processing
+===================
+
+The execution framework can automatically recover from failures, in the sense of restoring full processing capacity without compromising the framework's state.
+However, recovering the data processing itself is the responsibility of the code running on each worker.
+
+Visit processing can fail at any point, including within and between tasks, and while performing external I/O.
+Even within a single worker, the order of events is not fully deterministic (e.g., two snaps could be ingested and processed up to snap combination in parallel).
+However, the possible failures can be divided into a handful of cases:
+
+- Any failure before the end of APDB writing (which is atomic as of May 2023) has no consequences beyond that worker's local repository.
+  The repository may have ingested raws, or partial pipeline outputs.
+- A failure after APDB writing but before the end of ``DiaPipelineTask``, which contains it, is recoverable only if the task can account for the fact that a previous session already wrote to the APDB.
+- A failure after APDB writing but before or during alert distribution cannot be aborted without leaving the APDB inconsistent with the alert stream.
+  However, retrying is possible, and easy if duplicate alerts are acceptable (see :ref:`general-priorities`).
+- A failure after alert distribution but but before or during central repo sync may leave the latter missing calexps and difference images, which are supposed to be exported to the Rubin Science Platform.
+- A failure after central repo sync may interfere with internal metrics and monitoring, but has no consequences for science users.
+
+In general, the Rubin Observatory project has followed a policy of processing data at most once, not at least once.
+We propose to do the same thing in Prompt Processing, by attempting to retry processing only in the case of specific failure modes that we know to be recoverable.
+As a first approximation, this means that retries are allowed before writing to the APDB, but not afterward.
+Any failures that are not retried automatically can still be handled in next-day processing.
+
+
+.. _association:
+
+APDB and Source Association
+===========================
+
+The persistent nature of the APDB makes it difficult to retry processing runs that modify it.
+For example, if DiaSource and DiaObject IDs are deterministic functions of only their visit, then pipeline code might handle retries by testing for these IDs in the APDB, and ignoring or overwriting them.
```

Comment:
Shouldn't we already know how DiaSource and DiaObject IDs are generated?

Reply:
Is the generation algorithm set in stone? I'm trying to avoid justifying any decisions based on details that may change later.

Reply:
I guess things might be non-deterministic if we, for example, parallelized detection over sections of the detector image.  So this is fine.